### ISSUE TITLE
Upload progress via informational responses

### ISSUE THREAD

**Issue by Acconut** [httpwg/http-extensions#2291]
In a message to the mailing list (https://lists.w3.org/Archives/Public/ietf-http-wg/2022OctDec/0006.html), Austin Wright (@awwright) proposed to use informational responses to send information about the upload progress from the server to the client. During the Upload Creation Procedure and Upload Appending Procedure, the server could regularly send an informational response with the latest offset in the Upload-Offset header.

The value should reflect the amount of data that has been received and safely stored by the server. This way, the client knows what data has been successfully transmitted and is not needed for resumption anymore. Any resources attached for this data (e.g. memory buffers) can be released then. In contrast, If the client only monitors which,data it has sent, it cannot be sure if the server properly stored it yet.

There is a short example of this how communication could look like (note: it uses server-generated upload URLs, but it could also be done using client-generated tokens, of course):

```
:method: POST
:scheme: https
:authority: example.com
:path: /upload
content-length: 300
[content (300 bytes)]

:status: 104
location: https://example.com/upload/b530ce8ff

:status: 1xx
upload-offset: 100

:status: 1xx
upload-offset: 200

:status: 201
location: https://example.com/upload/b530ce8ff
upload-offset: 300
```

**Comment by kazuho** [comment]
I think doing this is fine, and I agree that it has benefits as pointed out above.

General guideline of HTTP is that servers are allowed to send any number of informational responses. There's no reason to not use that.

Regarding the status code, I think we can reuse 104.

**Comment by LPardue** [comment]
+1 Kazuho

**Comment by guoye-zhang** [comment]
I think we need some communication between client and server to enable this. The protocol is designed to avoid depending on 1xx status codes.

* Does the client need to explicitly opt in to this?
* Can the server opt out?
* What if the server doesn't support 1xx?

If we want to keep this optional without adding a new header field, a possible design is:

```
:status: 104
location: /upload/token
upload-offset: 0

:status: 104
upload-offset: 10000
```

If the server wants to send periodic updates, it can start by doing so in the initial 104. The client can enable additional buffering behavior based on this information (but it might have already dropped some bytes by the initial 104). This is a server driven solution and the client won't be able to opt out.

**Comment by reschke** [comment]
Or maybe https://greenbytes.de/tech/webdav/rfc2518.html#STATUS_102

**Comment by LPardue** [comment]
I'm not fussed on the specific response code value but the 102 seems deprecated, for whatever that means, and the linked wording seems at odds with the system design we want

**Comment by reschke** [comment]
It's not deprecated, but wasn't carried over to WebDAV-bis due to lack of implementations.

It's in the registry, so it's a real HTTP status code.

That said, I now agree it doesn't fit well, as 102 is for situations where the client is actually *waiting* for a response (as opposed to still sending).

**Comment by LPardue** [comment]
> It's not deprecated, but wasn't carried over to WebDAV-bis due to lack of implementations.

I guess that's accurate to the historical record, after a very cursory delve into RFC 2518 and 4918. But I wonder how clear that interpretation is to the broader HTTP world. MDN has a perspective you might want to follow up on.

> 
> It's in the registry, so it's a real HTTP status code.
> 
> That said, I now agree it doesn't fit well, as 102 is for situations where the client is actually _waiting_ for a response (as opposed to still sending).

Yeah this is where I'm hitting a hard edge. A registered status code, where the behaviour, the important stuff,  is described in an obsoleted RFC doesn't really seem great.

**Comment by reschke** [comment]
FWIW, I would push back hard on that line of argument. The code is registered, so it's a valid status code.

And yes, a spec that wants to *use* might want to either include an updated definition, or make sure the definition is moved to a separate new document.

**Comment by awwright** [comment]
I also think that 102 is OK to use here (but maybe a new status code is still warranted, see below). The 102 isn't depreciated as a bad idea or because of another better solution, it was merely omitted from recent updates to the WebDAV specification for lack of implementations. I imagine another specification could take over ownership of the definition/registration.

I think the only reason I didn't specifically write 102 in my email is, I've [previously (2019) experimented with using it](https://awwright.github.io/http-progress/draft-wright-http-progress.html) to indicate "upload complete, the server is now performing additional processing" and it may be useful to distinguish this signal from "upload has been partially received and committed to storage". The former is probably closer to the intended usage, so it may still be good to mint a new status code for the latter usage.

Alternatively, both cases might be distinguished by the HTTP fields that are returned, and maybe servers want to indicate both at the same time ("50% committed to disk, 20% processed").

**Comment by kazuho** [comment]
I share the view that 102 does not fit very well here. It would be awkward to send an URI to which the missing part of the request should be sent, while indicating that "the server has accepted the complete request" ([RFC 2158 section 10.1](https://datatracker.ietf.org/doc/html/rfc2518#section-10.1)).

**Comment by Acconut** [comment]
I opened a PR for this in https://github.com/httpwg/http-extensions/pull/2664.

**Comment by smatsson** [comment]
I think this would be a good addition to the spec as long as it's not mandatory for the server to send this data. As previously stated in this thread It would help the client to discard in-memory data that is no longer required to resume.

Would it make sense to also include the location header in the additional 104 responses? This would make the solution more robust in case one (or several) of the 104 responses get lost and the extra header field would not add any significant amount of extra bytes in each request.

**Comment by Acconut** [comment]
> I think this would be a good addition to the spec as long as it's not mandatory for the server to send this data.

Yes, this would be optional.




> Would it make sense to also include the location header in the additional 104 responses? This would make the solution more robust in case one (or several) of the 104 responses get lost and the extra header field would not add any significant amount of extra bytes in each request.

I am unsure how big the benefit would actually be. This would only help if a 104 is lost without the connection being destroyed, so the HTTP response is still open and a subsequent 104 could be received. Maybe this is possible with HTTP/2 or HTTP/3, but I don't know that. Other than that, repeating Location would be redundant.

**Comment by LPardue** [comment]
1xx response are allowed to be dropped by intermediaries. But I'm not aware of any that would drop _some_ in a request. Typically they just swallow all or pass all.

**Comment by kazuho** [comment]
@LPardue 
> 1xx response are allowed to be dropped by intermediaries. 

Is that so? I could be wrong but [RFC 9110 section 15.2](https://www.rfc-editor.org/rfc/rfc9110.html#name-informational-1xx) states that _a proxy MUST forward 1xx responses unless the proxy itself requested the generation of the 1xx response_.

**Comment by MikeBishop** [comment]
I suspect what @LPardue meant to say is that some proxies _do_ drop 1xx responses, whether they're allowed to or not. Also, some HTTP libraries consume all the 1xx responses without exposing them to the caller, even if the library itself doesn't do anything with that particular 1xx code; such a library is indistinguishable from a poorly-behaved intermediary.

**Comment by smatsson** [comment]
> I am unsure how big the benefit would actually be. This would only help if a 104 is lost without the connection being destroyed, so the HTTP response is still open and a subsequent 104 could be received. Maybe this is possible with HTTP/2 or HTTP/3, but I don't know that. Other than that, repeating Location would be redundant.

> 1xx response are allowed to be dropped by intermediaries. But I'm not aware of any that would drop some in a request. Typically they just swallow all or pass all.

My concern isn't so much proxies dropping one or more response as much as e.g. switching network and losing one or more of the 104 response (think spotty Wi-Fi moving over to 4G). Another input here would be consistency, especially if we add the `Upload-Offset: 0` header to the first 104. I agree that the benefit might not be huge but it's not zero either. It does however seem like the consensus in Prague was that this is not needed and I can get behind that.

**Comment by mnot** [comment]
> e.g. switching network and losing one or more of the 104 response (think spotty Wi-Fi moving over to 4G)

One or more of the 104 responses could be dropped, along with the final response, due to the connection being interrupted, but _selective_ 104s can't be dropped due to network issues. Clients need to deal with failed connections no matter what, so  I don't see how this is an issue.

**Comment by smatsson** [comment]
@mnot Right, then this is a non-issue that can be skipped. I just wanted to bring it up so that we do not miss aspects. :)

**Comment by Acconut** [comment]
Thanks for the feedback! I adjusted https://github.com/httpwg/http-extensions/pull/2664 to clarify what header fields can be present in 104 responses.
